### DISCUSSION TITLE
Proposal: Add Debouncing to `useQuery` Hook Using the `take` Option

### DISCUSSION THREAD

**Discussion by Amr2** [TanStack/query#8423] [Ideas] [answered]
## **Overview**
The purpose of this proposal is to introduce a **debounce mechanism** for the `useQuery` hook in React Query. This would provide developers with the ability to control how frequently queries are triggered when inputs change (like search inputs or dynamic form filters).

The key idea is to add a new **`take` option argument** to `useQuery`, which determines whether to trigger the **first, last, or every** occurrence of the query within a specified debounce window.

---

## **Problem Statement**
Currently, developers need to manually debounce queries when working with features like **typeahead search**, **auto-complete**, or **real-time filters**. This requires additional logic to be implemented by every developer, leading to repetitive, boilerplate code. 

While React Query offers **staleTime** and **cacheTime** options, they do not address the issue of input-driven debounce logic. Most developers rely on libraries like `lodash.debounce` or create custom wrappers, which adds unnecessary complexity to projects.

---

## **Proposed Solution**
We propose adding a **`take` option** to the `useQuery` configuration object, allowing developers to specify which event should "win" during a period of rapid consecutive calls.

### **New Option**
```javascript
const { data } = useQuery(queryKey, queryFn, {
  take: 'first' | 'last' | 'every', // New option (default: 'every')
  debounceTime: 300, // The debounce window (in milliseconds)
});
```
<h3><strong>Option Definitions</strong></h3>

Option | Description
-- | --
take: 'first' | Triggers the first query call in the debounce window and ignores subsequent triggers.
take: 'last' | Triggers only the last query call in the debounce window. This is the most common form of debouncing.
take: 'every' | Allows every call to go through after a debounce delay. This acts more like a rate limiter.
debounceTime | The debounce window in milliseconds (default: 0).

<blockquote>
<p><strong>Default behavior</strong>: If <code inline="">debounceTime</code> is not provided, no debounce logic is applied. If <code inline="">take</code> is not set, it defaults to <code inline="">'every'</code> to match the current behavior of <code inline="">useQuery</code>.</p>
</blockquote>
<hr>
<h2><strong>Usage Examples</strong></h2>
<h3><strong>1️⃣ Take the "Last" Event (Classic Debounce)</strong></h3>
<p>This is the most typical use case for debouncing — only trigger the last query after the user stops typing for a specified delay.</p>
<pre><code class="language-javascript">
const { data } = useQuery(['search', searchTerm], fetchSearchResults, {
  debounceTime: 300, 
  take: 'last', // Triggers only the last event after 300ms of inactivity
});
</code></pre>
<p><strong>Use case</strong>: Autocomplete or search bars where the last input is the only one that matters.</p>
<hr>
<h3><strong>2️⃣ Take the "First" Event</strong></h3>
<p>This approach is useful when you want to <strong>capture the first change</strong> and ignore everything else that happens within the debounce window.</p>
<pre><code class="language-javascript">const { data } = useQuery(['search', searchTerm], fetchSearchResults, {
  debounceTime: 300, 
  take: 'first', // Triggers only the first query within the 300ms debounce window
});
</code></pre>
<p><strong>Use case</strong>: For situations like form submission, where you want to react immediately on the first action and ignore subsequent rapid actions.</p>
<hr>
<h3><strong>3️⃣ Take "Every" Event (Rate Limiting)</strong></h3>
<p>This allows every event to be processed but only after a <strong>minimum debounce delay</strong> between queries.</p>
<pre><code class="language-javascript">const { data } = useQuery(['search', searchTerm], fetchSearchResults, {
  debounceTime: 300, 
  take: 'every', // Triggers every event after a 300ms delay
});
</code></pre>
<p><strong>Use case</strong>: Ideal for scenarios where you want to send periodic updates to an API but still reduce the frequency of requests.</p>
<hr>
<h2><strong>Proposed Implementation</strong></h2>
<h3><strong>Implementation Steps</strong></h3>
<ol>
<li><strong>Extend the <code inline="">useQuery</code> Hook</strong>:
<ul>
<li>Add the <code inline="">take</code> and <code inline="">debounceTime</code> options to the hook's parameter configuration.</li>
</ul>
</li>
<li><strong>Internal Debounce Logic</strong>:
<ul>
<li>Use an internal <strong>debounce queue</strong> for each query key.</li>
<li>Track the <strong>first, last, and all</strong> trigger times.</li>
</ul>
</li>
<li><strong>Update the Query Execution Logic</strong>:
<ul>
<li>Apply logic to track which event to "take" (first, last, every) during the debounce period.</li>
<li>Use a mechanism similar to <code inline="">setTimeout</code> or <strong>requestAnimationFrame</strong> to clear pending calls and trigger only the required one.</li>
</ul>
</li>
</ol>
<hr>
<h2><strong>Alternatives Considered</strong></h2>
<ol>
<li><strong>Custom Wrapper Hooks</strong>: Developers can create custom hooks for debounce logic, but this leads to duplicated logic across projects.</li>
<li><strong>Global Debounce Option</strong>: Add <code inline="">debounceTime</code> to the global query client configuration. However, since debounce needs to be contextual, it makes more sense to keep it at the query level.</li>
<li><strong>Leave It Out</strong>: This leaves developers relying on libraries like <code inline="">lodash.debounce</code>, but this increases the learning curve for new users.</li>
</ol>
<hr>
<h2><strong>Benefits of This Proposal</strong></h2>
<ul>
<li><strong>Improves DX</strong>: No need to create custom debounce logic in every project.</li>
<li><strong>Reduces Boilerplate</strong>: Avoids importing <code inline="">lodash.debounce</code> and wrapping hooks.</li>
<li><strong>Enhanced Performance</strong>: Reduces unnecessary API requests.</li>
<li><strong>Backward Compatible</strong>: Defaults for <code inline="">take: 'every'</code> and <code inline="">debounceTime: 0</code> ensure that existing projects won't break.</li>
</ul>
<hr>
<h2><strong>Drawbacks</strong></h2>
<ul>
<li><strong>Complexity</strong>: Adds additional logic to the <code inline="">useQuery</code> hook.</li>
<li><strong>Redundancy</strong>: Some may argue that <code inline="">staleTime</code> and <code inline="">cacheTime</code> already provide sufficient control, but <strong>they do not solve user-input-driven debounce issues</strong>.</li>
</ul>
<hr>
<h2><strong>Request for Feedback</strong></h2>
<p>We’d love feedback on the following aspects:</p>
<ol>
<li><strong>API Design</strong>: Does the proposed <code inline="">take: 'first' | 'last' | 'every'</code> make sense, or would you suggest a different API?</li>
<li><strong>Use Cases</strong>: Are there use cases we missed where this logic would be valuable?</li>
<li><strong>Implementation Approach</strong>: Would you suggest changes to make it more performant or simpler?</li>
</ol>
<hr>
<p>We look forward to your thoughts and feedback!</p>
</body></html>
</body>
</html>

**Comment by TkDodo** [comment, marked as answer]
Thanks for the detailed proposal - it warrants a detailed answer.

We’ve stated in the past that `debouncing` is one feature that will _not_ make it into the Query API. I’ve talked about that in my recent talk at the React Advanced Conference “React Query API Design - Lessons Learned”:

-  [slides](https://tkdodo.eu/blog/react-query-api-design-lessons-learned)
-  [recording](https://gitnation.com/contents/react-query-api-design-lessons-learned)

The reason is manifold:

1. It adds complexity to the API

`lodash.debounce` has a lot of options - `leading`, `trailing`, `maxWait`, which I think is what users would expect us to support, too. Some users might want to have `throttling` instead - another option we’d need to support. All of this adds to bundle size as well - `lodash.debounce` is 1.14kb, which is roughly a 10% increase of our current size.

2. It’s easily doable in user-land

You’re saying it “adds unnecessary complexity to projects”, I would argue the opposite that it keeps unnecessary complexity out of React Query. It’s  really just one additional line of code:

```diff
+  const debouncedSearchTerm = useDebounce(searchTerm, 500)

  const { data } = useQuery({
-   queryKey: ['search', searchTerm]
+   queryKey: ['search', debouncedSearchTerm]
    queryFn: fetchSearchResults,
  });
```

You can pick your favourite `useDebounce` hook that debounces a value, or write one yourself.

3. User-land can actually do it better!

With your proposal, the QueryKey would change with every keystroke, which creates a new cache entry with every keystroke, and then we would somehow debounce the call to the QueryFn. The user-land implementation debounces the queryKey, which means React Query will only know about the new cache entry after debouncing has finished. This is conceptually much closer to what you’d expect, as we wouldn’t create that many in-between entries.

4. Debouncing itself is likely a workaround

For example, for auto-complete, we likely just don’t want to send events until the user has “stopped” typing. Defining an arbitrary time for what “stopped” means seems ... arbitrary? In my example, it will also delay the fetch by 500ms after the user has stopped.

Since React 18, React offers the built-in `useDeferredValue` hook for low-prio updates that can be interrupted by high-prio updates (e.g. user input). With the user-land solution, you can do:

```js
const deferredSearchTerm = React.useDeferredValue(searchTerm)

const { data } = useQuery({
  queryKey: ['search', deferredSearchTerm]
  queryFn: fetchSearchResults,
});
```

and kind of get the best of both worlds with a react built-in feature.

**Reply by KevinVandy** [reply]
I'll add to this here by saying that we're building TanStack Pacer to make doing debouncing/throttling etc. easier in userland for query or anything else. Our goal is to adds a handful of examples for how to integrate TanStack Pacer with TanStack Query.

https://tanstack.com/pacer/latest
